src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt

```diff
@@ -1,124 +1,198 @@
+        .betweenInside(IDENTIFIER, ENUM_DISCRIMINANT, ENUM_VARIANT).spaces(1)
+
+        //== types
+        .afterInside(LIFETIME, REF_TYPE).spaceIf(true)
+        .betweenInside(ts(MUL), ts(CONST, MUT), PTR_TYPE).spaces(0)
+        .before(TYPE_PARAM_BOUNDS).spaces(0)
+        .beforeInside(LPAREN, PATH_PART).spaces(0)
+
+        //== expressions
+        .beforeInside(LPAREN, PAT_ENUM).spaces(0)
+        .beforeInside(LBRACK, INDEX_EXPR).spaces(0)
+        .afterInside(PARAMS_LIKE, LAMBDA_EXPR).spacing(1, 1, 0, true, 1)
+
+        //== macros
+        .betweenInside(IDENTIFIER, EXCL, MACRO_INVOCATION).spaces(0)
+        .between(MACRO_INVOCATION, MACRO_ARGS).spaces(0)
+        .betweenInside(MACRO_INVOCATION, IDENTIFIER, MACRO_DEFINITION).spaces(1)
+        .betweenInside(IDENTIFIER, MACRO_ARG, MACRO_DEFINITION).spaces(1)
+
+        //== rules with very large area of application
+        .around(NO_SPACE_AROUND_OPS).spaces(0)
+        .around(SPACE_AROUND_OPS).spaces(1)
+        .around(KEYWORDS).spaces(1)
+
+fun computeSpacing(parentBlock: Block, child1: Block?, child2: Block, ctx: RustBlockContext): Spacing? {
+    if (child1 is AbstractRustBlock && child2 is AbstractRustBlock) {
+        val origNode1 = child1.node;
+        val origNode2 = child2.node;
+        val origPsi1 = origNode1.psi
+        val origPsi2 = origNode2.psi
+
+        // Here are rules which are comment-sensitive
+        if (origPsi1 is RustOuterAttr && (origPsi2 is RustOuterAttr || origPsi1.parent is RustItem)) {
+            return lineBreak(keepBlankLines = 0)
+        }
+
+        // Handle blocks of comments to get proper spacing between items and statements
+        val (psi1, psi2) = omitComments(origNode1, origPsi1, origNode2, origPsi2)
+
+        // Here are rules which are comment-insensitive
+        if (psi1 is RustStmt && psi2.isStmtOrExpr) {
+            return lineBreak(
+                keepLineBreaks = ctx.commonSettings.KEEP_LINE_BREAKS,
+                keepBlankLines = ctx.commonSettings.KEEP_BLANK_LINES_IN_CODE)
+        }
+
+        if (psi1.isTopLevelItem && psi2.isTopLevelItem) {
+            return lineBreak(
+                keepLineBreaks = ctx.commonSettings.KEEP_LINE_BREAKS,
+                keepBlankLines = ctx.commonSettings.KEEP_BLANK_LINES_IN_DECLARATIONS)
         }
-        return self
     }
+    return ctx.spacingBuilder.getSpacing(parentBlock, child1, child2)
 }
+
+private fun omitComments(origNode1: ASTNode, origPsi1: PsiElement,
+                         origNode2: ASTNode, origPsi2: PsiElement): Pair<PsiElement, PsiElement> =
+    Pair(
+        if (origPsi1 is PsiComment && origNode1.hasLineBreakAfterInSameParent) {
+            origPsi1.getPrevNonCommentSibling() ?: origPsi1
+        } else {
+            origPsi1
+        },
+        if (origPsi2 is PsiComment && origNode2.hasLineBreakBeforeInSameParent) {
+            origPsi2.getNextNonCommentSibling() ?: origPsi2
+        } else {
+            origPsi2
+        }
+    )
```

Comment:
Ugh comments between formatting blocks are very burdensome. I hope the heuristics I have used here will work (almost) flawlessly.